cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java

```diff
@@ -28,55 +27,47 @@
   private final Map<String, String> properties;
   /**
    * Sequence id of operations on the preferences
+   * Normally this should be > 0. But can be 0 indicating that no preferences have been set on the entity.
    */
   private final Long seqId;
   /**
    * Whether it is a resolved preferences or not.
    */
   private boolean resolved;
 
-  public static PreferencesDetail merge(PreferencesDetail left, PreferencesDetail right) {
-    if (left == null && right == null) {
-      return null;
-    } else if (left == null) {
-      return right;
-    } else if (right == null) {
-      return left;
-    }
-
+  /**
+   * Return a resolved preference detail where preferences in {@code parent} take precedence over
+   * those in {@code child}. The {@code seqId} would be the max of the two.
+   */
+  public static PreferencesDetail resolve(PreferencesDetail parent, PreferencesDetail child) {
     Map<String, String> properties = new HashMap<>();
-    properties.putAll(left.getProperties());
-    properties.putAll(right.getProperties());
+    // Copy child's properties first.
+    properties.putAll(child.getProperties());
+    // Add parent's properties and overrides any existing properties in child;
+    properties.putAll(parent.getProperties());
 
-    Long seqId = null;
-    Long leftSeqId = left.getSeqId();
-    Long rightSeqId = right.getSeqId();
-    if (leftSeqId != null && rightSeqId != null) {
-      seqId = Long.max(leftSeqId.longValue(), rightSeqId.longValue());
-    } else if (leftSeqId != null) {
-      seqId = new Long(leftSeqId.longValue());
-    } else if (rightSeqId != null) {
-      seqId = new Long(rightSeqId.longValue());
-    }
+    Long seqId = getMaxSeqId(parent.getSeqId(), child.getSeqId());
 
-    boolean resolved = (left.resolved || right.resolved);
-    return new PreferencesDetail(properties, seqId, resolved);
+    return new PreferencesDetail(properties, seqId, true);
   }
 
-  public PreferencesDetail(Map<String, String> properties, @Nullable Long seqId, boolean resolved) {
+  public PreferencesDetail(Map<String, String> properties, Long seqId, boolean resolved) {
```

Comment:
if it can't be null, it should be a long instead of a Long

Reply:
Done.